src/framework/rendering/Loader.cpp

```diff
@@ -0,0 +1,234 @@
+                                   const vector<vec3>& texcoords, const vector<Triangle>& triangle, vector<vec4>& tangent)
+{
+   vec3 *tan1 = new vec3[vertexCount * 2];
+   vec3 *tan2 = tan1 + vertexCount;
+   ZeroMemory(tan1, vertexCount * sizeof(vec3) * 2);
+
+   for(long a = 0; a < triangle.size(); a++) {
+      long i1 = triangle[a].index[0];
+      long i2 = triangle[a].index[1];
+      long i3 = triangle[a].index[2];
+
+      const vec3& v1 = vertices[i1];
+      const vec3& v2 = vertices[i2];
+      const vec3& v3 = vertices[i3];
+
+      const vec2& w1 = vec2(texcoords[i1]);
+      const vec2& w2 = vec2(texcoords[i2]);
+      const vec2& w3 = vec2(texcoords[i3]);
+
+      float x1 = v2.x - v1.x;
+      float x2 = v3.x - v1.x;
+      float y1 = v2.y - v1.y;
+      float y2 = v3.y - v1.y;
+      float z1 = v2.z - v1.z;
+      float z2 = v3.z - v1.z;
+
+      float s1 = w2.x - w1.x;
+      float s2 = w3.x - w1.x;
+      float t1 = w2.y - w1.y;
+      float t2 = w3.y - w1.y;
+
+      float r = 1.0F / (s1 * t2 - s2 * t1);
+      vec3 sdir((t2 * x1 - t1 * x2) * r, (t2 * y1 - t1 * y2) * r, (t2 * z1 - t1 * z2) * r);
+      vec3 tdir((s1 * x2 - s2 * x1) * r, (s1 * y2 - s2 * y1) * r, (s1 * z2 - s2 * z1) * r);
+
+      tan1[i1] += sdir;
+      tan1[i2] += sdir;
+      tan1[i3] += sdir;
+
+      tan2[i1] += tdir;
+      tan2[i2] += tdir;
+      tan2[i3] += tdir;
+   }
+
+   for(long a = 0; a < vertexCount; a++) {
+      const vec3& n = normals[a];
+      const vec3& t = tan1[a];
+
+      // Gram-Schmidt orthogonalize
+      tangent.push_back(vec4(glm::normalize(t - n * glm::dot(n, t)), 1.0f));
+
+      // Calculate handedness
+      tangent[a].w = (glm::dot(glm::cross(n, t), tan2[a]) < 0.0F) ? -1.0F : 1.0F;
+   }
+
+   delete[] tan1;
+}
+
+//---------------------------------------------------------------------------------------------------
+bool Loader::LoadMesh(const char* filePath, MeshData& meshData, bool calculateTangents)
+{
+   meshData.Clear();
+
+   int iRetVal;
+   vector<vec3> tmpVertices, tmpNormals, tmpTexCoords;
+   vector<int> vertexIndices, normalIndices, textureIndices;
+
+   FILE* fp;
+   fopen_s(&fp, filePath, "r");
+   if(fp == NULL) {
+      return false;
+   }
+
+   vec3 v, n, tc;
+
+   char c;
+   while(1) {
+      c = fgetc(fp);
+      if(c == EOF) {
+         break;
+      }
+      else if(c == '\n')
+         continue;
+
+      if(c == 'v') {
+         c = fgetc(fp);
+         if(c == 'n') {
+            iRetVal = fscanf_s(fp, " %f %f %f\n", &n.x, &n.y, &n.z);
+            tmpNormals.push_back(n);
+         }
+         else if(c == 't') {
+            iRetVal = fscanf_s(fp, " %f %f %f\n", &tc.x, &tc.y, &tc.z);
+            tmpTexCoords.push_back(tc);
+         }
+         else {
+            iRetVal = fscanf_s(fp, " %f %f %f\n", &v.x, &v.y, &v.z);
+            tmpVertices.push_back(v);
+         }
+      }
+      else if(c == 'f') {

```

Comment:
I think we need to change this since this seems to be not handling faces having more than 3 vertices.